pkg/queue/request_metric.go

```diff
@@ -17,166 +17,75 @@ limitations under the License.
-	if err != nil {
-		return nil, err
-	}
-
-	return &requestMetricsHandler{
-		next:     next,
-		statsCtx: ctx,
-	}, nil
-}
-
-func (h *requestMetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	rr := pkghttp.NewResponseRecorder(w, http.StatusOK)
-	startTime := time.Now()
-
-	defer func() {
-		// Filter probe requests for revision metrics.
-		if netheader.IsProbe(r) {
-			return
-		}
-
-		// If ServeHTTP panics, recover, record the failure and panic again.
-		err := recover()
-		latency := time.Since(startTime)
-		routeTag := GetRouteTagNameFromRequest(r)
-		if err != nil {
-			ctx := metrics.AugmentWithResponseAndRouteTag(h.statsCtx,
-				http.StatusInternalServerError, routeTag)
-			pkgmetrics.RecordBatch(ctx, requestCountM.M(1),
-				responseTimeInMsecM.M(float64(latency.Milliseconds())))
-			panic(err)
-		}
-		ctx := metrics.AugmentWithResponseAndRouteTag(h.statsCtx,
-			rr.ResponseCode, routeTag)
-		pkgmetrics.RecordBatch(ctx, requestCountM.M(1),
-			responseTimeInMsecM.M(float64(latency.Milliseconds())))
-	}()
-
-	h.next.ServeHTTP(rr, r)
+	duration metric.Float64Histogram
+	queueLen metric.Int64Gauge
 }
 
 // NewAppRequestMetricsHandler creates an http.Handler that emits request metrics.
-func NewAppRequestMetricsHandler(next http.Handler, b *Breaker,
-	ns, service, config, rev, pod string,
+func NewAppRequestMetricsHandler(
+	mp metric.MeterProvider,
+	next http.Handler,
+	b *Breaker,
 ) (http.Handler, error) {
-	keys := []tag.Key{metrics.PodKey, metrics.ContainerKey, metrics.ResponseCodeKey, metrics.ResponseCodeClassKey}
-	if err := pkgmetrics.RegisterResourceView(&view.View{
-		Description: "The number of requests that are routed to user-container",
-		Measure:     appRequestCountM,
-		Aggregation: view.Count(),
-		TagKeys:     keys,
-	}, &view.View{
-		Description: "The response time in millisecond",
-		Measure:     appResponseTimeInMsecM,
-		Aggregation: defaultLatencyDistribution,
-		TagKeys:     keys,
-	}, &view.View{
-		Description: "The number of items queued at this queue proxy.",
-		Measure:     queueDepthM,
-		Aggregation: view.LastValue(),
-		TagKeys:     keys,
-	}); err != nil {
-		return nil, err
+	var (
+		meter   = mp.Meter(scopeName)
+		err     error
+		handler = &appRequestMetricsHandler{
+			next:    next,
+			breaker: b,
+			clock:   clock.RealClock{},
+		}
+	)
+
+	handler.duration, err = meter.Float64Histogram(
+		"kn.queueproxy.app.duration",
+		metric.WithDescription("The duration of task execution"),
+		metric.WithUnit("s"),
+		metric.WithExplicitBucketBoundaries(latencyBounds...),
+	)
+	if err != nil {
+		panic(err)
 	}
 
-	ctx, err := metrics.PodRevisionContext(pod, "queue-proxy", ns, service, config, rev)
+	handler.queueLen, err = meter.Int64Gauge(
+		"kn.queueproxy.depth",
+		metric.WithDescription("Number of current items in the queue"),
+		metric.WithUnit("{item}"),
+	)
 	if err != nil {
-		return nil, err
+		panic(err)
```

Comment:
is there a reason we are switching from returning an error to panicking on an error?